### PR TITLE
bpo-31714: Improved regular expression documentation.

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue31714

### CODE REVIEW

File: Doc/howto/regex.rst [hunk 1]
```diff
@@ -549,27 +545,39 @@ of each one.
 
    Perform case-insensitive matching; character class and literal strings will
    match letters by ignoring case.  For example, ``[A-Z]`` will match lowercase
-   letters, too, and ``Spam`` will match ``Spam``, ``spam``, or ``spAM``. This
-   lowercasing doesn't take the current locale into account; it will if you also
-   set the :const:`LOCALE` flag.
+   letters, too. Full Unicode matching also works unless the :const:`re.ASCII`
+   flag is also used to disable non-ASCII matches.  ``[A-Z]`` will also match
```
Comment (ezio-melotti) [inline Doc/howto/regex.rst]: ``disable non-ASCII matches`` makes it sounds like it won't match any non-ASCII characters, but that is false:
```
>>> re.match('负鼠', '负鼠', re.ASCII)
<_sre.SRE_Match object; span=(0, 2), match='负鼠'>
```
I think it would be more correct to say that regex sets will only match characters in the ASCII range.

File: Doc/howto/regex.rst [hunk 2]
```diff
@@ -549,27 +545,39 @@ of each one.
...
+   letters 'İ' (U+0130, Latin capital letter I with dot above), 'ı' (U+0131,
+   Latin small letter dotless i), 'ſ' (U+017f, Latin small letter long s) and
+   'K' (U+212a, Kelvin sign) in Unicode mode.  ``Spam`` will match ``Spam``,
```
Comment (ezio-melotti) [inline Doc/howto/regex.rst]: I think this is wrong: unless I'm mistaken ``[A-Z]`` should be limited to upper case ASCII letters, even with re.UNICODE.
Perhaps you meant to use ``\w``?
It would also be better to specify what Unicode categories are matched, instead of just providing a few examples and letting the user figure it out from there.
I think this is already explained below, so a link or a mention to that section is fine.

File: Doc/howto/regex.rst [hunk 3]
```diff
@@ -549,27 +545,39 @@ of each one.
...
-   your system is configured properly and a French locale is selected, certain C
-   functions will tell the program that ``'é'`` should also be considered a letter.
+   Make ``\w``, ``\W``, ``\b``, ``\B`` and case-insensitive matching dependent
+   on the current locale instead of the Unicode database.
+
+   Locales are a feature of the C library intended to help in writing programs
+   that take account of language differences.  For example, if you're
+   processing encoded French text, you'd want to be able to write ``\w+`` to
+   match words, but ``\w`` only matches the character class ``[A-Za-z]`` in
+   bytes patterns; it won't match bytes corresponding to ``'é'`` or ``'ç'``.
```
Comment (ezio-melotti) [inline Doc/howto/regex.rst]: I would remove the ``'...'`` here and perhaps add ``b'\xe9'`` and ``b'\xe7'`` within parenthesis.

File: Doc/howto/regex.rst [hunk 4]
```diff
@@ -549,27 +545,39 @@ of each one.
...
+   If your system is configured properly and a French locale is selected,
+   certain C functions will tell the program that the byte corresponding
+   ``'é'`` should also be considered a letter.
```
Comment (ezio-melotti) [inline Doc/howto/regex.rst]: corresponding _to_ ``é``
I also don't like ``certain C functions will tell the program`` too much.

File: Doc/howto/regex.rst [hunk 5]
```diff
@@ -549,27 +545,39 @@ of each one.
...
    Setting the :const:`LOCALE` flag when compiling a regular expression will cause
    the resulting compiled object to use these C functions for ``\w``; this is
    slower, but also enables ``\w+`` to match French words as you'd expect.
+   The use of this flag is discouraged in Python 3 as the locale mechanism
+   is very unreliable, and it only handles one "culture" at a time anyway;
+   and it only works with 8-bit locales;
+   you should use Unicode matching instead, which is the default in Python 3
+   for Unicode (str) patterns.
```
Comment (ezio-melotti) [inline Doc/howto/regex.rst]: ... is very unreliable, it only handles one "culture" at a time, and it only works with 8-bit locales.  You should use ...

Since Unicode matching is the default, I wouldn't say "you should use", but just something like "Unicode matching is already enabled by default in Python 3, and it is able to handle different locales/languages."

File: Doc/howto/regex.rst [hunk 6]
```diff
@@ -667,8 +675,8 @@ zero-width assertions should never be repeated, because if they match once at a
 given location, they can obviously be matched an infinite number of times.
 
 ``|``
-   Alternation, or the "or" operator.   If A and B are regular expressions,
-   ``A|B`` will match any string that matches either ``A`` or ``B``. ``|`` has very
+   Alternation, or the "or" operator.   If *A* and *B* are regular expressions,
+   ``A|B`` will match any string that matches either *A* or *B*. ``|`` has very
```
Comment (ezio-melotti) [inline Doc/howto/regex.rst:681]: Why `*` instead of ` `` `?

File: Doc/library/re.rst [hunk 7]
```diff
@@ -537,26 +544,39 @@ form.
...
           IGNORECASE
 
    Perform case-insensitive matching; expressions like ``[A-Z]`` will also
-   match lowercase letters.  The current locale does not change the effect of
-   this flag.  Full Unicode matching (such as ``Ü`` matching ``ü``) also
-   works unless the :const:`re.ASCII` flag is also used to disable non-ASCII
-   matches.
-
+   match lowercase letters.  Full Unicode matching (such as ``Ü`` matching
+   ``ü``) also works unless the :const:`re.ASCII` flag is also used to disable
```
Comment (ezio-melotti) [inline Doc/library/re.rst]: is ~~also~~ used to
(you already have one before)

File: Doc/library/re.rst [hunk 8]
```diff
@@ -537,26 +544,39 @@ form.
...
+   non-ASCII matches.  The current locale does not change the effect of this
+   flag unless the :const:`re.LOCALE` flag is also used.
+   Correcsponds the inline flag ``(?i)``.
+
+   Note that in Unicode patterns the expression ``[A-Z]`` will match not
+   only letters 'A' to 'Z' and 'a' to 'z', but will also match letters 'İ'
+   (U+0130, Latin capital letter I with dot above), 'ı' (U+0131, Latin small
+   letter dotless i), 'ſ' (U+017f, Latin small letter long s) and 'K' (U+212a,
+   Kelvin sign).  If the :const:`ASCII` flag is used, only letters 'a' to 'z'
```
Comment (ezio-melotti) [inline Doc/library/re.rst]: This is duplicated from above, so the same comments applied (maybe it shouldn't be duplicated?).

File: Doc/library/re.rst [hunk 9]
```diff
@@ -537,26 +544,39 @@ form.
...
-   current locale. The use of this flag is discouraged as the locale mechanism
+   Make ``\w``, ``\W``, ``\b``, ``\B`` and case-insensitive matching
+   dependent on the current locale.  This flag can be used only with bytes
+   patterns.  The use of this flag is discouraged as the locale mechanism
    is very unreliable, and it only handles one "culture" at a time anyway;
+   and it only works with 8-bit locales;
    you should use Unicode matching instead, which is the default in Python 3
-   for Unicode (str) patterns. This flag can be used only with bytes patterns.
+   for Unicode (str) patterns.
+   Correcsponds the inline flag ``(?L)``.
```
Comment (ezio-melotti) [inline Doc/library/re.rst]: Corre~~c~~sponds _to_
(extra c and missing to)

Also applies below.

File: Doc/library/re.rst [hunk 10]
```diff
@@ -809,6 +831,13 @@ form.
       >>> print('|'.join(map(re.escape, sorted(operators, reverse=True))))
       /|\-|\+|\*\*|\*
 
+   This functions must not be used for the replacement string in :func:`sub`
+   and :func:`subn`, only backslashes should be escaped.  For example::
+
+      >>> int_re = r'\d+'
+      >>> print(re.sub('INT', int_re.replace('\\', r'\\'), r'INT(\.INT)?(e[+-]?INT)?'))
+      \d+(\.\d+)?(e[+-]?\d+)?
```
Comment (ezio-melotti) [inline Doc/library/re.rst]: I don't find this example particularly clear.  Why would someone want to use re.escape() on the replacement string?  Wouldn't using ``int_re = r'\\d+'`` (and a normal str.replace on INT) be easier?

**Review by ezio-melotti — CHANGES_REQUESTED** [review]
I added a few comments.
There are also a couple of paragraphs with long lines that you could rewrap.

File: Doc/howto/regex.rst [hunk 11]
```diff
@@ -1284,12 +1292,12 @@ doesn't work because of the greedy nature of ``.*``. ::
    >>> print(re.match('<.*>', s).group())
    <html><head><title>Title</title>
 
-The RE matches the ``'<'`` in ``<html>``, and the ``.*`` consumes the rest of
+The RE matches the ``<`` in ``'<html>'``, and the ``.*`` consumes the rest of
```
Comment (cryvate) [inline Doc/howto/regex.rst]: I think you do want ``'<'`` as in the original here. See also line 1299 in this commit.

File: Doc/howto/regex.rst [hunk 12]
```diff
@@ -667,8 +675,8 @@ zero-width assertions should never be repeated, because if they match once at a
 given location, they can obviously be matched an infinite number of times.
 
 ``|``
-   Alternation, or the "or" operator.   If A and B are regular expressions,
-   ``A|B`` will match any string that matches either ``A`` or ``B``. ``|`` has very
+   Alternation, or the "or" operator.   If *A* and *B* are regular expressions,
+   ``A|B`` will match any string that matches either *A* or *B*. ``|`` has very
    low precedence in order to make it work reasonably when you're alternating
    multi-character strings. ``Crow|Servo`` will match either ``Crow`` or ``Servo``,
```
Comment (cryvate) [inline Doc/howto/regex.rst]: everything after "will match either" have ``'[word]'`` instead of ``[word]``?

File: Doc/howto/regex.rst (hunk 2 above)
Comment (cryvate) [inline Doc/howto/regex.rst]: shouldn't all those after "will match" be ``'[word]'`` instead of ``[word]``?

File: Doc/howto/regex.rst [hunk 13]
```diff
@@ -229,24 +226,23 @@ Another repeating metacharacter is ``+``, which matches one or more times.  Pay
 careful attention to the difference between ``*`` and ``+``; ``*`` matches
 *zero* or more times, so whatever's being repeated may not be present at all,
 while ``+`` requires at least *one* occurrence.  To use a similar example,
-``ca+t`` will match ``cat`` (1 ``a``), ``caaat`` (3 ``a``'s), but won't match
-``ct``.
+``ca+t`` will match ``'cat'`` (1 ``'a'``), ``'caaat'`` (3 ``a``'s), but won't
```
Comment (cryvate) [inline Doc/howto/regex.rst]: (3 ``'a'``s)?

File: Doc/howto/regex.rst (hunk 1 above)
Comment (serhiy-storchaka) [reply Doc/howto/regex.rst]: This was a copy from existing re module documentation. :(

> regex sets will only match characters in the ASCII range.

This doesn't sound good too. Regex sets can match characters outsides the ASCII range with the re.ASCII flag.

>>> re.match('[耀-鿐]+', '负鼠', re.ASCII)
<re.Match object; span=(0, 2), match='负鼠'>

But case-insensitive matching works only in the ASCII range. 'é' doesn't match 'É' with the re.ASCII flag.

File: Doc/howto/regex.rst (hunk 2 above)
Comment (serhiy-storchaka) [reply Doc/howto/regex.rst]: Not that this is about case-insensitive matching. 'S' matches both 's' and 'ſ'.

Comment (serhiy-storchaka) [reply Doc/howto/regex.rst]: Sorry, I'm not sure that I understood your comment.

File: Doc/howto/regex.rst (hunk 3 above)
Comment (serhiy-storchaka) [reply Doc/howto/regex.rst]: It is not clear why b'\xe9' and b'\xe7' should be matched. But 'é' and 'ç' are French letters, and I have added "bytes corresponding to" for making this phrase Python 3 compatible.

File: Doc/howto/regex.rst (hunk 4 above)
Comment (serhiy-storchaka) [reply Doc/howto/regex.rst]: Do you suggest remove quotes?

`certain C functions will tell the program` already was here, it looks correct to me, and I don't know how improve it.

File: Doc/howto/regex.rst (hunk 6 above)
Comment (serhiy-storchaka) [reply Doc/howto/regex.rst:681]: Because A and B are not literal regexpes, but variables.

File: Doc/library/re.rst (hunk 8 above)
Comment (serhiy-storchaka) [reply Doc/library/re.rst]: Could you suggest how to avoid duplication?

File: Doc/library/re.rst (hunk 10 above)
Comment (serhiy-storchaka) [reply Doc/library/re.rst]: I tried to provide simplified real-word example. In Mailman re.sub() is used for creating a regular expression. They passed the pattern containing a \d as a replacement string and got an error when this became invalid. Someone could use re.escape() on the replacement string, because the replacement string looks similar to simple pattern (it expands \n and \1). And this will work while the replacement string don't contain other metacharacters except a backslash.

I'll replace this example with the better one.

**Review by serhiy-storchaka — COMMENTED** [review]
Thank you @ezio-melotti and @cryvate! I'll rewrap long lines after resolving all issues with wording.

File: Doc/howto/regex.rst (hunk 3 above)
Comment (ezio-melotti) [reply Doc/howto/regex.rst]: The reason why I suggested that, is because ``'é'`` and ``'ç'`` are Unicode str in Python 3, whereas without quotes they are just letters.  The addition of ``b'\xe9'`` and ``b'\xe7'`` might help clarify what is being matched, but it's not essential.

File: Doc/library/re.rst (hunk 8 above)
Comment (ezio-melotti) [reply Doc/library/re.rst]: I didn't notice the duplicated part was on two separate files, so it's probably ok to leave it.

File: Doc/howto/regex.rst [hunk 14]
```diff
@@ -526,7 +522,7 @@ of each one.
 +=================================+============================================+
 | :const:`ASCII`, :const:`A`      | Makes several escapes like ``\w``, ``\b``, |
 |                                 | ``\s`` and ``\d`` match only on ASCII      |
-|                                 | characters with the respective property.   |
+|                                 | characters with the respective property    |
```
Comment (warsaw) [inline Doc/howto/regex.rst]: I guess this is done for consistency?  Any reason these shouldn't be complete sentences?  I'd have rather added missing periods instead of removed them.  Does this make the diff smaller?  Is that worth it?

**Review by warsaw — COMMENTED** [review]
Looks pretty good to me.  I have only one dumb comment; feel free to ignore it :)

File: Doc/howto/regex.rst (hunk 14 above)
Comment (serhiy-storchaka) [reply Doc/howto/regex.rst]: No other reasons besides consistency.

File: Doc/howto/regex.rst (hunk 2 above)
Comment (cryvate) [reply Doc/howto/regex.rst]: ``Spam``, ``spam``, ``spAM``, or ``ſpam`` ->  ``'Spam'``, ``'spam'``, ``'spAM'``, or ``'ſpam'``? I might have gotten it wrong though.

Comment (ezio-melotti) [reply Doc/howto/regex.rst]: I did some tests:
```
>>> unichars = ''.join(chr(cp) for cp in range(0x110000))
>>> ''.join(re.findall('[a-z]', bmp))
'abcdefghijklmnopqrstuvwxyz'
>>> ''.join(re.findall('[A-Z]', bmp))
'ABCDEFGHIJKLMNOPQRSTUVWXYZ'
>>> ''.join(re.findall('[a-z]', bmp, re.I))
'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyzİıſK'
>>> ''.join(re.findall('[A-Z]', bmp, re.I))
'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyzİıſK'
```
Now I understand what the paragraph means: [a-z] and [A-Z] will only match ASCII lower and upper case letters respectively (so 26 chars each), however if re.I is used with either one, since those 4 letters (and only those 4) are valid capitalization of the 26 ASCII letters, they will be matched as well (bringing the total up to 26 + 26 + 4 == 56).
Do you think we should rephrase it to make it clearer?

File: Doc/howto/regex.rst (hunk 14 above)
Comment (cryvate) [reply Doc/howto/regex.rst]: If it is single sentences in a table, I would normally not have a period, but that's just me.

File: Doc/howto/regex.rst (hunk 2 above)
Comment (ezio-melotti) [reply Doc/howto/regex.rst]: I think he's referring to the 4 non-ASCII letters: they should be enclosed in  `` `'..'` ``

File: Doc/howto/regex.rst (hunk 4 above)
Comment (ezio-melotti) [reply Doc/howto/regex.rst]: I suggest to remove quotes, since we are talking about the character, not about a Unicode string.
Fair enough about the wording being there already.

File: Doc/howto/regex.rst (hunk 2 above)
Comment (serhiy-storchaka) [reply Doc/howto/regex.rst]: It would be very good to make it clearer. Now, since you understand what the paragraph means, could you please suggest a clear wording?

This example is not artificial. See [bpo-31672](https://bugs.python.org/issue31672). I think this caveat should be documented specially.

Comment (ezio-melotti) [reply Doc/howto/regex.rst]:
```
When the patterns ``[a-z]`` or ``[A-Z]`` are used in combination with the re.I and re.U flag,
they will match the 52 ASCII letters and 4 additional non-ASCII letters: 'İ' (U+0130, Latin
capital letter I with dot above), 'ı' (U+0131, Latin small letter dotless i), 'ſ' (U+017F,
Latin small letter long s) and 'K' (U+212A, Kelvin sign).
```

**Comment by serhiy-storchaka** [comment]
I didn't expect the Spanish Inquisition! (Neither Italian)

**Comment by ezio-melotti** [comment]
There is still (at least) a comment that is not addressed, but GitHub decided to hide it.  You can see by clicking the second "show outdated" link from the top, or by checking the latest emails for this issue to make sure  you don't miss any comment.

**Comment by serhiy-storchaka** [comment]
Oh, sorry, @ezio-melotti, I missed this comment. Please check whether I understood you correctly.

**Review by ezio-melotti — APPROVED** [review]

**Comment by serhiy-storchaka** [comment]
Thank you all for your reviews and suggestions!
